### PR TITLE
AGENT-5439 download remote artifacts prior to NeMo server start

### PR DESCRIPTION
**Pull request by alexromanyuk**
## Changes
`Triton Predictor`:
- added calculation of the `Inference-Header-Content-Length` header, which is sent by client but removed from the request by the Prediction API `predictUnstructured` route. This header is required by Triton [Binary Tensor server extension](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/protocol/extension_binary_data.html).

`Triton Drop-in environment`:
- set the `model_config` source files to `PYTHONPATH`. The `model_config` is generated from the corresponding [Triton version protobuf schema (r24.02)](https://github.com/triton-inference-server/common/blob/r24.02/protobuf/model_config.proto). It's used to parse model's `config.pbtxt` configuration.
- updated the dr_requirements.txt

`Nemo Predictor`:
- added support of custom model `load_model` hook, since it's not available for language predictors out of the box.
- ~~added support of the `ModelDownloader` tool, which searches for the `.remote` file artifacts and then downloads models~~ not yet. this library should be reviewed and published separately.

`Nemo Drop-in environment`
- ~~added the [datarobot_custom_code source files](https://github.com/datarobot/datarobot-custom-code) to the `PYTHONPATH`. We're going to release this code as a separate library, but until then we need it to be in the Nemo environment. This library is used to download model files in parallel and report status into MLOPS Health.~~ not yet. this library should be reviewed and published separately.
- updated the dr_requirements.txt 

## Testing
All the changes were tested manually. I'm going to cover predictors / environments with the functional tests in the upcoming PRs.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -131,3 +131,6 @@ tests/functional/custom_java_predictor/**/*.jar
 # misc
 .metals
 .vscode
+
+# used to download large LLM models, library is not published yet
+!datarobot_custom_code-0.0.1-py3-none-any.whl
```
Comment (alexromanyuk) [inline .gitignore:136]: github repo [is here](https://github.com/datarobot/datarobot-custom-code)

Comment (alexromanyuk) [reply .gitignore:136]: @elatt , it's a separate repo and It's not really ready to be published as a library anywhere.
So I built a wheel and included it into the NIM drop-in environment.

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py [hunk 2]
```diff
@@ -94,6 +113,54 @@ def supported_payload_formats(self):
...
+                "The following custom model files were found:\n"
+            )
+            error_mes += "\n".join([str(path) for path in (custom_py_paths + dot_remote_paths)])
+            self.logger.error(error_mes)
+            raise DrumCommonException(error_mes)
+
+        return custom_py_paths, dot_remote_paths
+
+    def _load_model_remote_artifact(self):
+        from datarobot_custom_code.model_artifact_downloader import ModelDownloader
```
Comment (alexromanyuk) [inline custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:134]: @elatt
this import is somehow interferes with the NIM start. It downloads all artifacts successfully, but then NIM can't be started. If I start NIM with the uploaded model directly, it starts just fine.

Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:134]: perhaps I need to deactivate environment properly, currently I just removing the `VIRTUAL_ENV/bin:` from the `PATH`.

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py [hunk 3]
```diff
@@ -94,6 +113,54 @@ def supported_payload_formats(self):
...
+            code_dir=self._code_dir,
+            chunk_size=1024 * 1024 * 10,
+            nr_processes=1,
+        )
+
+        self.logger.info("Downloading remote files")
+        download_info = md.download_remote_files()
+        self.logger.info(md.prepare_download_summary(download_info))
+
+    def _load_model_with_custom_hooks(self, params):
```
Comment (alexromanyuk) [inline custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:148]: @elatt this is how Language Predictors load/use hooks

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py (hunk 2 above)
Comment (elatt) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:134]: Yeah, my first guess would be with venv (I haven't looked in a long time what all the magic `deactivate` shell function does). You can probably just update your subprocess call to `shell=True` and then do `deactivate && <cmd to run NIM>` right?

Comment (elatt) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:134]: Actually there isn't that much magic
```sh
deactivate () {
    # reset old environment variables
    if [ -n "${_OLD_VIRTUAL_PATH:-}" ] ; then
        PATH="${_OLD_VIRTUAL_PATH:-}"
        export PATH
        unset _OLD_VIRTUAL_PATH
    fi
    if [ -n "${_OLD_VIRTUAL_PYTHONHOME:-}" ] ; then
        PYTHONHOME="${_OLD_VIRTUAL_PYTHONHOME:-}"
        export PYTHONHOME
        unset _OLD_VIRTUAL_PYTHONHOME
    fi

    # Call hash to forget past commands. Without forgetting
    # past commands the $PATH changes we made may not be respected
    hash -r 2> /dev/null

    if [ -n "${_OLD_VIRTUAL_PS1:-}" ] ; then
        PS1="${_OLD_VIRTUAL_PS1:-}"
        export PS1
        unset _OLD_VIRTUAL_PS1
    fi

    unset VIRTUAL_ENV
    unset VIRTUAL_ENV_PROMPT
    if [ ! "${1:-}" = "nondestructive" ] ; then
    # Self destruct!
        unset -f deactivate
    fi
}
```
- Cleanup `PATH`
- Cleanup `PYTHONHOME` (only relevant if base image had PYTHONHOME set for some reason) 
- cleanup `PS1` (not relevant for your script)
- cleanup some other venv related env vars (not sure but also probably not relevant)
- run `hash` (pretty sure not relevant when we run a command via `subprocess` with `shell=False`)

So, in summary, I think as long as you are fixing up `PATH` correctly, I think things should work.

File: .gitignore (hunk 1 above)
Comment (elatt) [reply .gitignore:136]: We can't merge this branch like this though.

Comment (elatt) [reply .gitignore:136]: Maybe like Trition, we need to do two phases, checkin first w/o this tool and have follow on work to add it later.

File: public_dropin_gpu_environments/nemollm_inference_ms/datarobot_custom_code-0.0.1-py3-none-any.whl
Comment (elatt) [inline public_dropin_gpu_environments/nemollm_inference_ms/datarobot_custom_code-0.0.1-py3-none-any.whl:1]: We can't checkin this file like this.

File: .gitignore (hunk 1 above)
Comment (alexromanyuk) [reply .gitignore:136]: Removed the wheel library, I will add the source code into the dropin environment instead.

File: public_dropin_gpu_environments/triton_server/start_server.sh [hunk 4]
```diff
@@ -21,7 +21,7 @@ fi
 echo
 echo "Executing command: tritonserver --model-repository=${CODE_DIR}"
 echo
-nohup tritonserver --model-repository=${MODEL_DIR} &
+nohup tritonserver --model-repository=${CODE_DIR}/model_repository &
```
Comment (alexromanyuk) [inline public_dropin_gpu_environments/triton_server/start_server.sh:24]: the model-repository path is error-prone:
- Nemo names it as `model-store`
- Triton names it as `model_repository`
- Custom model tends to drop high level directory without any notice.

This should be fixed in a separate PR: users directory structure should be reconstructed, so Triton can read it. If it's not possible - raise an error.

File: .gitignore (hunk 1 above)
Comment (elatt) [reply .gitignore]: Yeah I think this is better. Just make this code part of drum and we can split it out if others want to use it.

**Comment by alexromanyuk** [comment]
No changes to review by the Core Modeling team. Please add yourself if you think otherwise :bow:

File: public_dropin_gpu_environments/nemollm_inference_ms/Dockerfile [hunk 5]
```diff
@@ -78,6 +78,11 @@ ENV PIP_NO_CACHE_DIR=1
 ENV PATH=${BASE_PATH}
 RUN python -m venv $DATAROBOT_VENV_PATH && $DATAROBOT_VENV_PATH/bin/pip install -U pip
 COPY dr_requirements.txt dr_requirements.txt
+
+# artifact downloader tool
+COPY artifact_downloader/ ./artifact_downloader/
+ENV PYTHONPATH=$PYTHONPATH:${CODE_DIR}/artifact_downloader/
```
Comment (alexromanyuk) [inline public_dropin_gpu_environments/nemollm_inference_ms/Dockerfile:84]: I moved the ModelDownloader code to a separate PR:
1) the code change is big
2) I'm not the original owner of this code, but we want to maintain and extend it
https://github.com/datarobot/datarobot-user-models/pull/965/files

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py [hunk 6]
```diff
@@ -94,6 +113,54 @@ def supported_payload_formats(self):
...
+            error_mes += "\n".join([str(path) for path in (custom_py_paths + dot_remote_paths)])
+            self.logger.error(error_mes)
+            raise DrumCommonException(error_mes)
+
+        return custom_py_paths, dot_remote_paths
+
+    def _load_model_remote_artifact(self):
+        from datarobot_custom_code.model_artifact_downloader import ModelDownloader
+
+        deployment_id = os.environ["MLOPS_DEPLOYMENT_ID"]
```
Comment (elatt) [inline custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:136]: This won't exist during custom models tests. Won't this break?

Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:136]: correct, it should not be in the CM tests. I'll update it

File: custom_model_runner/datarobot_drum/drum/enum.py [hunk 7]
```diff
@@ -17,6 +17,7 @@
 
 PRED_COLUMN = "Predictions"
 
+DOT_REMOTE = ".remote"
```
[the remark this answers is not in the document]
Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/enum.py:20]: will change it to the `REMOTE_FILE_EXT`

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py [hunk 8]
```diff
@@ -95,8 +98,29 @@ def has_read_input_data_hook(self):
...
+
+        # Predictions API does not forward the headers,
+        # while Triton expects to get inference header for the Binary Tensor request. See
+        # https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/protocol/extension_binary_data.html
+        if INFERENCE_HEADER.lower() not in set(k.lower() for k in headers):
+            inference_header_size = self._get_inference_header_size(data)
+            if inference_header_size:
+                headers.update({INFERENCE_HEADER: str(inference_header_size)})
+
+        self.logger.info(headers)
```
Comment (elatt) [inline custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py:123]:
```suggestion
        self.logger.debug(headers)
```
This looks more like debgging info?

File: public_dropin_gpu_environments/nemollm_inference_ms/Dockerfile (hunk 5 above)
Comment (alexromanyuk) [reply public_dropin_gpu_environments/nemollm_inference_ms/Dockerfile:84]: I decided to remove the ModelDownloader from this release, until we fix all the issues related to it.

Comment (alexromanyuk) [reply public_dropin_gpu_environments/nemollm_inference_ms/Dockerfile:84]: will remove those lines

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py (hunk 6 above)
Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py]: Removed this library at all. I'll create a separate tickets to make:
- tests
- harness pipeline
- fix logging and other misc things

File: public_dropin_gpu_environments/nemollm_inference_ms/Dockerfile (hunk 5 above)
Comment (alexromanyuk) [reply public_dropin_gpu_environments/nemollm_inference_ms/Dockerfile]: removed.

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py (hunk 8 above)
Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py]: updated

File: public_dropin_gpu_environments/triton_server/model_config/README.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+This folder contains the Triton model configuration parser generated from the protocol buffer scheme (r24.02)
```
Comment (elatt) [inline public_dropin_gpu_environments/triton_server/model_config/README.md:1]: Is it really 24.02? I thought we had to revert back to 24.01 due to Onnx issue?

File: public_dropin_gpu_environments/triton_server/dr_requirements.txt [hunk 10]
```diff
@@ -1,2 +1,140 @@
...
+    #   pandas
+    #   strictyaml
+pytz==2024.1
+    # via pandas
+pyyaml==6.0.1
+    # via
+    #   datarobot
+    #   datarobot-drum
 requests==2.31.0
-datarobot-drum==1.10.21
```
Comment (elatt) [inline public_dropin_gpu_environments/triton_server/dr_requirements.txt:2]: I guess we are going back to v1.10.20?

File: public_dropin_gpu_environments/triton_server/dr_requirements.txt [hunk 11]
```diff
@@ -1,2 +1,140 @@
+#
+# This file is autogenerated by pip-compile with Python 3.9
+# by the following command:
+#
+#    pip-compile --index-url=https://artifactory.devinfra.drdev.io/artifactory/api/pypi/pypi/simple/ --no-emit-index-url --no-emit-trusted-host dr_requirements.in
```
Comment (elatt) [inline public_dropin_gpu_environments/triton_server/dr_requirements.txt:5]: We must not expose our internal artifactory URL.

File: public_dropin_gpu_environments/nemollm_inference_ms/dr_requirements.txt [hunk 12]
```diff
@@ -4,6 +4,10 @@
 #
 #    pip-compile --index-url=https://artifactory.devinfra.drdev.io/artifactory/api/pypi/pypi/simple/ --no-emit-index-url --no-emit-trusted-host dr_requirements.in
```
Comment (elatt) [inline public_dropin_gpu_environments/nemollm_inference_ms/dr_requirements.txt:5]: must not expose our internal artifactory

File: public_dropin_gpu_environments/triton_server/model_config/README.md (hunk 9 above)
Comment (alexromanyuk) [reply public_dropin_gpu_environments/triton_server/model_config/README.md:1]: yep. I took scheme from the 24.02, but it works with 24.01 w/o issues

File: public_dropin_gpu_environments/triton_server/dr_requirements.txt (hunk 10 above)
Comment (alexromanyuk) [reply public_dropin_gpu_environments/triton_server/dr_requirements.txt:2]: the current drum version is [v.1.10.21](https://github.com/datarobot/datarobot-user-models/blob/master/custom_model_runner/datarobot_drum/drum/description.py#L7).

I'll need to bump version and update dropin env afterwards.

File: public_dropin_gpu_environments/triton_server/dr_requirements.txt (hunk 11 above)
Comment (alexromanyuk) [reply public_dropin_gpu_environments/triton_server/dr_requirements.txt]: removed

File: public_dropin_gpu_environments/nemollm_inference_ms/dr_requirements.txt (hunk 12 above)
Comment (alexromanyuk) [reply public_dropin_gpu_environments/nemollm_inference_ms/dr_requirements.txt]: removed

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py [hunk 13]
```diff
@@ -95,8 +98,29 @@ def has_read_input_data_hook(self):
     def _predict(self, **kwargs) -> RawPredictResponse:
         raise DrumCommonException("Predict structured is not supported")
 
+    def _get_inference_header_size(self, data):
+        try:
+            first_line = io.BytesIO(data).readline()
+            json.loads(first_line)  # read first line to check it's a valid json string
+            return len(first_line)
```
Comment (elatt) [inline custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py:105]: What's in this JSON payload, can there be Unicode characters? I guess it doesn't matter if it does or not? Also does `first_line` contain the `\n` character?

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py [hunk 14]
```diff
@@ -94,6 +115,41 @@ def supported_payload_formats(self):
...
+
+    def _load_model_with_custom_hooks(self, params):
+        try:
+            model_adapter = PythonModelAdapter(
+                model_dir=self._code_dir, target_type=self.target_type
+            )
+            sys.path.append(self._code_dir)
+            model_adapter.load_custom_hooks()
+            # disable the check of scoring hook existence
+            model_adapter._custom_hooks[CustomHooks.SCORE] = True
```
Comment (zohar-mizrahi) [inline custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:145]: I don't see the full picture, but the expected value is a method rather than a boolean. Please validate that it does not break things up.

**Review by zohar-mizrahi — APPROVED** [review]
Given the comment about the missing tests, I assume you'll add it later.
LGTM

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py (hunk 14 above)
Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py:145]: validated, this hook is never invoked

File: custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py (hunk 13 above)
Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py:105]: > What's in this JSON payload
example payload:
```
POST /v2/models/mymodel/infer HTTP/1.1
Host: localhost:8000
Content-Type: application/octet-stream
Inference-Header-Content-Length: <xx>
Content-Length: <xx+19>
{
  "model_name" : "mymodel",
  "inputs" : [
    {
      "name" : "input0",
      "shape" : [ 2, 2 ],
      "datatype" : "UINT32",
      "parameters" : {
        "binary_data_size" : 16
      }
    },
    {
      "name" : "input1",
      "shape" : [ 3 ],
      "datatype" : "BOOL",
      "parameters" : {
        "binary_data_size" : 3
      }
    }
  ],
  "outputs" : [
    {
      "name" : "output0",
      "parameters" : {
        "binary_data" : true
      }
    }
  ]
}
<16 bytes of data for input0 tensor>
<3 bytes of data for input1 tensor>
```
https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/protocol/extension_binary_data.html#examples

Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py:105]: > can there be Unicode characters

yes, unicode characters can be in the name of input and output layers.

Comment (alexromanyuk) [reply custom_model_runner/datarobot_drum/drum/gpu_predictors/triton_predictor.py:105]: > Also does first_line contain the \n character

yes, the `first_line` must contain the `\n` character in order we can parse it. I added the example client code to the https://github.com/datarobot/datarobot-user-models/pull/966/files
